patterns-use-cases/payment-state-machine/payment-state-machine-java/src/main/java/my/example/PaymentService.java

```diff
@@ -0,0 +1,122 @@
+
+package my.example;
+
+import dev.restate.sdk.ObjectContext;
+import dev.restate.sdk.annotation.Handler;
+import dev.restate.sdk.annotation.VirtualObject;
+import dev.restate.sdk.common.StateKey;
+import dev.restate.sdk.common.TerminalException;
+import dev.restate.sdk.serde.jackson.JacksonSerdes;
+import my.example.accounts.AccountClient;
+import my.example.types.Payment;
+import my.example.types.PaymentStatus;
+import my.example.types.Result;
+import my.example.utils.TypeChecks;
+
+import java.time.Duration;
+
+/**
+ * The service that processes the payment requests.
+ *
+ * This is implemented as a virtual object to ensure that only
+ * one concurrent request can happen per token (requests are queued and
+ * processed sequentially per token).
+ *
+ * Note that this idempotency-token is more of an operation/payment-id.
+ * Methods can be called multiple times with the same token, but payment
+ * will be executed only once. Also, if a cancellation is triggered for that
+ * token, the payment will not happen or be undine, regardless of whether
+ * the cancel call comes before or after the payment call.
+ */
+@VirtualObject
+public class PaymentService {
+
+  private final long EXPIRY_TIMEOUT = 60 * 60 * 1000L; // 1 hour
+
+  private static final StateKey<PaymentStatus> STATUS =
+          StateKey.of("status", JacksonSerdes.of(PaymentStatus.class));
+  private static final StateKey<Payment> PAYMENT =
+          StateKey.of("payment", JacksonSerdes.of(Payment.class));
+
+  @Handler
+  public Result makePayment(ObjectContext ctx, Payment payment) {
+    // De-duplication to make calls idempotent
+    PaymentStatus status = ctx.get(STATUS).orElse(PaymentStatus.NEW);
+    if (status == PaymentStatus.CANCELLED) {
+        return new Result(false, "Payment already cancelled");
+    }
+
+    if (status == PaymentStatus.COMPLETED_SUCCESSFULLY) {
+        return new Result(false, "Payment already completed in prior call");
+    }
+
+    // We check the types, because external calls via http may evade some type checks
+    TypeChecks.validatePayment(payment);
+
+    // Charge the target account
+    Result paymentResult = AccountClient.fromContext(ctx, payment.getAccountId())
+            .withdraw(payment.getAmountCents()).await();
+
+    // Remember only on success, so that on failure (when we didn't charge) the external
+    // caller may retry this (with the same token), for the sake of this example
+    if (paymentResult.isSuccess()) {
+        ctx.set(STATUS, PaymentStatus.COMPLETED_SUCCESSFULLY);
+        ctx.set(PAYMENT, payment);
+
+        String idempotencyToken = ctx.key();
+        PaymentServiceClient.fromContext(ctx, idempotencyToken)
+                .send(Duration.ofMillis(EXPIRY_TIMEOUT))
+                .expireToken();
+    }
+
+    return paymentResult;
+  }
+
+  @Handler
+  public void cancelPayment(ObjectContext ctx) {
+    PaymentStatus status = ctx.get(STATUS).orElse(PaymentStatus.NEW);
+
+    switch (status){
+      case NEW:
```

Comment:
You can use the new switch statement here https://www.baeldung.com/java-switch#1-the-new-switch-expression